### DISCUSSION TITLE
Complete Current Statement on Enter breaks Rename

### DISCUSSION THREAD

**Discussion by chylex** [JetBrains/ideavim#397] [Q&A] [answered]
Hi. I have the following binding:

```
nmap <Enter> :action EditorCompleteStatement<CR>
```

However, if I use the Rename refactoring and for ex. delete a character in normal mode, then pressing Enter runs the action before the Rename finishes, which discards all the changes I made while renaming.

Is there a way to ensure the mapping does not run while renaming an element? Thanks.

**Comment by AlexPl292** [comment]
Hi, that's an interesting case. Actually, I think it's possible using the vim script, so you could create some function like
```
function! EnterFunction()
  if (renaming)
    <Enter>
  else
    :action EditorCompleteStatement<CR>
  endif
endfunction

nmap <expr> <Enter> EnterFunction()
```
However, at the moment this is not possible to verify if the renaming is in progress. We'll discuss the possibility of adding such function.

**Reply by chylex** [reply]
Turns out that doesn't quite work out of the box; if the `if` statement resolves into `<Enter>` then it seems IdeaVim tries to insert a new line instead of confirming the refactoring, so I get this error.

```
2021-11-01 20:09:37,774 [ 154629]   WARN - .maddyhome.idea.vim.KeyHandler - Caught exception during CompleteStatement()
Must not change document outside command or undo-transparent action. See com.intellij.openapi.command.WriteCommandAction or com.intellij.openapi.command.CommandProcessor 
```

I'll investigate further.

**Reply by chylex** [reply]
```
function! CompleteStatement()
  if (renaming())
    <Enter>
  else
    :action EditorCompleteStatement<CR>
  endif
endfunction
```

Actually, the `<Enter>` branch gets resolved to `ShiftLeftCommand(ranges=Ranges[ranges=[]], argument=>, length=1)` instead of the Enter action. Not sure if this is a known bug or maybe already fixed, my version is a few days old.

EDIT: Didn't work, it also gets confused about `<CR>` and thinks it's part of the action name so it cannot find the action. If I remove `<CR>`, it finds the action, but doesn't indent the caret correctly.

**Reply by lippfi** [reply]
It is not a bug. `<Enter>` in a separate line is fairly recognized as a ShiftLeftCommand with trailing characters. And `<CR>` at the end of `action EditorCompleteStatement<CR>` is also wrong. Every line should be a command, not a sequence of keys to press in normal mode.
Map with expression evaluates passed expression (`CompleteStatement()` in our case) and uses its value as for regular mapping, so the proper vim script function should look like this:
```
function! CompleteStatement()
  if (renaming())
    return "\<Enter>"
  else
    action EditorCompleteStatement
    return ""
  endif
endfunction
```
But there is a known bug with `\<Enter>` and `<Enter>`: for vim the first is an escaped enter key and 
 the second is a sequence of characters (see `:h string`), but in IdeaVim behavior is opposite (I will fix it in 1.9 or 1.10 release)

**Reply by chylex** [reply]
Thanks, got it to work. Want me to PR the `renaming` function?

**Reply by AlexPl292** [reply]
Have you implemented `renaming` function by yourself? Yes, we can check your implementation, however we haven't yet considered the naming convention for functions that don't exist in vim.

Sorry about the incorrect script, this was actually a draft from the where I've tried to describe the general idea. I haven't checked if this is actually work.

**Reply by chylex** [reply]
https://github.com/chylex/IntelliJ-IdeaVim/commit/04fe85dd7969d75f6d419a2902251d27c91c3acb

**Reply by chylex** [reply]
Hi, would you like me to PR the `renaming` function, and is there any update on the convention for functions that don't exist in vim?

**Reply by lippfi** [reply]
Hi!
Regarding the naming convention for Vim script functions:
I believe the best option would be to start all functions that aren't native to Vim with a capital letter. This approach will help us avoid naming conflicts with Vim in the future.

As for the PR:
The function is platform-dependent and might require maintenance in the future due to potential API updates. I doubt its widespread usage, not because the function is useless, but due to the current scarcity of other Vim script functions in IdeaVim. I don't foresee this situation changing in the next couple of years. If we decide to merge the PR, we'll also need a strategy to inform users about this new function so they're aware of its existence.

As I mentioned earlier, this platform-dependent code isn't something users would include in their `.vimrc` anyway. At present, I believe the optimal approach would be to create a Kotlin plugin for the IJ Platform with the desired functionality and then setting up mappings in `.ideavimrc`. While I don't believe many users directly accessing the `rename()` function, I can envision a complex plugin that leverage it for more advanced features. I'm optimistic about the future of Kotlin plugins for IdeaVim as opposed to an complex Vim script intertwined with IDE-specific features. I would like to make Kotlin plugins for IdeaVim easier to develop and use, though there aren't concrete plans yet.

But this is only my opinion. @AlexPl292 , what do you think?

**Reply by chylex** [reply]
I could see Kotlin plugins being a good approach, but that's assuming IdeaVim would have their own way of loading them.

JetBrains IDEs' plugin management is frankly not good; if I had to create an IDE plugin, either publish it to the marketplace (slow and too much admin work) or a custom plugin repository (not well supported), and then deal with IDE plugin installation and updates constantly breaking - then repeat for every IdeaVIM feature I'd like - I would rather just put it all into my own fork of IdeaVIM like I do now, and deal with having just one plugin to publish and manage in my IDE instances.

If `.ideavimrc` had syntax to load an IdeaVIM plugin from a `.jar` or `.kts` file, that would be great though.

**Reply by AlexPl292** [reply]
Hi there. Thank you for the proposal, but I think we won't merge it now because there are a bunch of open questions, and there is a single known use of this function. As for questions, we don't yet have the rules for the names of functions that are IdeaVim specific, and we're not completely sure how they should be implemented. Also, the new remote dev brings some complications around using the "smart" functionality of the IDE, and we don't yet have any tools to make sure that this function works properly there.
However, this function seems simple enough to be merged later.

**Reply by chylex** [reply]
There might be a better solution for this specific use case, such as not allowing IdeaVIM to take over Enter (or the configured IDEA keybinding for accepting the change) when a refactoring action is in progress. Is there any use case where someone might legitimately want IdeaVIM to interfere with the keybinding that confirms a refactoring?

**Reply by chylex** [reply]
I implemented this solution in my fork, it seems to make the original mapping `nmap <Enter> :action EditorCompleteStatement<CR>` work well: https://github.com/chylex/IntelliJ-IdeaVim/commit/080dbe9c0eebb5eb80f3a189cf656fdbe7aca331

(edited, found a bug)

**Reply by chylex** [reply]
I've been using the solution in https://github.com/chylex/IntelliJ-IdeaVim/commit/080dbe9c0eebb5eb80f3a189cf656fdbe7aca331 for a while now without noticing any issues. Would you accept a PR for it?

**Reply by AlexPl292** [reply, marked as answer]
Hi there! We now have a bit of an unstable state around enters and escapes. We migrated to a different way of handling them, but it slowly turned out that the previous one seemed to be the correct one. So, I'd like to stabilize things before doing further changes around key processing.

BTW, with the latest version of IdeaVim, your described case works for me without any fixes. However, as there is a chance that we'll get back to the previous solution, this problem may arise again.
I've created a ticket and I'll add it to the scope of the enter key questions. So, we'll create a UI test for it, and I'll check if works with the different implementations of key processing. If the issue exists, let's discuss your solution.

[VIM-3254](https://youtrack.jetbrains.com/issue/VIM-3254)
